### PR TITLE
Regression 2

### PR DESCRIPTION
**Pull request by ndoshi83**

### CODE REVIEW

**Comment by ndoshi83** [comment]
Assignment 1

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Nayan
## Assignment: Log_Linear_Regression


# What to work on
- [x] Finish your assignment
- [x] Show up to stand up!
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Nayan Doshi
## Sprint Challenge: Regression 2 


# Part 1. Wrangle Data

***
**What went well:**

-Nice merge and use of pandas to merge the two datasets:

```python
# Merge store df to train and test
train_m = pd.merge(train, store, on='Store')
test_m = pd.merge(test, store, on='Store')

train_m.shape, test_m.shape

```

- Nice plot distribution using matplotlib: 

```python 
# Plot distribution of target before and after logs
%matplotlib inline
plt.figure(figsize=(20, 30))
y_train.plot.hist()
y_train_log.plot.hist()
plt.title('Store Sales vs Log of Store Sales')

```

**What could be better**


- Plot could be easily applied using seaborn:

```python 
%matplotlib inline
import seaborn as sns
sns.distplot(y_train);


sns.distplot(y_train_log);

```



# Fit and validate your model

- Good cross val score using xgboost and category encoding: 

```python
import category_encoders as ce
from xgboost import XGBRegressor
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

encoder = ce.OrdinalEncoder()
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)
eval_set = [(X_train_encoded, y_train_log),
           (X_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log,
         eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)
```

**What could be better**

- Making a pipeline with k_means and XGBRegressor is an ideal solution as well (slightly better score):

```python
%%time
import category_encoders as ce
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import make_pipeline
#from conda install scikit-learnxgboost import XGBRegressor

pipeline = make_pipeline(ce.OrdinalEncoder(), XGBRegressor(random_state=42, n_jobs=-1))

k = 3
groups = train['Store']
scores = cross_val_score(pipeline, X_train, y_train_log, cv=k, n_jobs=-1, 
                         scoring='neg_mean_squared_error', groups=groups)
print(f'RMSLE for {k} folds:', np.sqrt(-scores))

```


# Plot model interpretation visualizations

- I like the output and the clean code! :

```python
import eli5
from eli5.sklearn import PermutationImportance

permuter = PermutationImportance(model, scoring='neg_mean_squared_error',
                                 cv ='prefit', n_iter=2, random_state=25)

permuter.fit(X_test_encoded, y_test_log)
feature_names = X_val_encoded.columns.tolist()
eli5.show_weights(permuter, top = None, feature_names = feature_names)
)

gs.fit(X_train, y_train_log);

print('best hyperparamters', search.best_params_)
print('cross val rmsle', np.sqrt(-search.best_score_))

#best hyperparamters {'xgbregressor__max_depth': 5, 'xgbregressor__n_estimators': 545}
#cross val rmsle 0.31088637924275075


```
**What could be better**

- pdp is a great way to see optimal category:

```python
from pdpbox.pdp import pdp_isolate, pdp_plot

feature = 'Customers'

isolated = pdp_isolate(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    feature=feature, 
    num_grid_points=20 
)

pdp_plot(isolated, feature_name=feature);
```


- Another example:

```python
from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['Customers', 'CompetitionDistance']

interaction = pdp_interact(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);
```


# What to work on
- [x] Try to read code backwards (makes for less complicated coding)
- [x] Come to PM for questions or help
